subsys/net/ip/tcp2.c

```diff
@@ -58,38 +58,57 @@ static size_t tcp_endpoint_len(sa_family_t af)
 		sizeof(struct sockaddr_in6);
 }
 
-static union tcp_endpoint *tcp_endpoint_new(struct net_pkt *pkt, int src)
+static union tcp_endpoint *tcp_endpoint_new(struct net_pkt *pkt)
 {
 	sa_family_t af = net_pkt_family(pkt);
 	union tcp_endpoint *ep = tcp_calloc(1, tcp_endpoint_len(af));
 
-	ep->sa.sa_family = af;
+	if (ep) {
+		ep->sa.sa_family = af;
+	}
 
-	switch (af) {
-	case AF_INET: {
-		struct net_ipv4_hdr *ip = (struct net_ipv4_hdr *)
-			net_pkt_ip_data(pkt);
-		struct tcphdr *th = th_get(pkt);
+	return ep;
+}
 
-		ep->sin.sin_port = src ? th->th_sport : th->th_dport;
+static union tcp_endpoint *tcp_endpoint_set(union tcp_endpoint *ep,
```

Comment:
I would suggest to change the return type to bool. This way it would be easier to do the error handling more clear and intuitive inside this function and in places of its usage.

Reply:
I did not want to change the current behaviour too much. So new `tcp_endpoint_set()` can be put to same place where `tcp_endpoint_new()` was earlier. Actually we do not seem to handle the out-of-mem case very well in current code either. Hmm, I will send another PR that adds more error checks, will check also would it be better to change the return type.

Reply:
Sure, this can be done later. These (handling out or memory and more explicit/clear error handling) were my points.

Alternative approach to consider later is just to embed endpoints into struct tcp and to drop the heap allocations.

Reply:
> Alternative approach to consider later is just to embed endpoints into struct tcp and to drop the heap allocations.

Yes, that would be better way indeed.